File: MergeSort.c
#include <stdio.h>

#define MAX 50

void merge(int arr[], int left, int mid, int right)
{
    int n1, n2;
    int i, j, k;

    n1 = mid - left + 1;
    n2 = right - mid;

    int L[n1], R[n2];

    for(i = 0; i < n1; ++i)
        L[i] = arr[left + i];

    for(j = 0; j < n2; ++j)
        R[j] = arr[mid + 1 + j];

    i = 0;
    j = 0;
    k = left;

    while(i < n1 && j < n2)
    {
        if(L[i] <= R[j])
            arr[k] = L[i++];
        else
            arr[k] = R[j++];
        k++;
    }

    while(i < n1)
        arr[k++] = L[i++];

    while(j < n2)
        arr[k++] = R[j++];
}

void MergeSort(int arr[], int left, int right)          
{
    int mid;

    if(left < right)
    {
        mid = left + (right - left)/2;
        MergeSort(arr, left, mid);
        MergeSort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }
}

void Display(int arr[], int arrSize)
{
    int i;

    for(i = 0; i < arrSize; ++i)
        printf("%d ", arr[i]);

    printf("\n");
}

int main()
{
    int arr[MAX];
    int size;
    int i;

    printf("Enter the size of array(max %d): ", MAX);
    scanf("%d", &size);

    if(size >= MAX)
    {
        printf("\nEntered size is greater than or equal to %d\n", MAX);
        return 0;
    }

    printf("\nEnter the elements in the array: ");

    for(i = 0; i < size; ++i)
        scanf("%d", &arr[i]);

    printf("\n---MERGE SORT---\n");
    
    MergeSort(arr, 0, size - 1);

    printf("\nElements in ascending order:\n");
    Display(arr, size);

    return 0;
}

File: Merge_Sort/MergeSort.java
import java.util.Scanner;

public class MergeSort
{
   static void merge(int arr[] , int beg , int mid , int end)
    {
        int nl = mid - beg +1 , nr = end-mid;
        int leftArr[] = new int[nl];
        int rightArr[] = new int[nr];

        for (int i = 0; i < nl; i++) {
            leftArr[i] = arr[beg+i];
        }
        for (int i = 0; i < nr; i++) {
            rightArr[i] = arr[mid +1 + i];
        }

        int i=0 , j=0 , k=beg;
        while(i<nl && j<nr){
            if(leftArr[i] <= rightArr[j])
            {
                arr[k]=leftArr[i];
                i++;
                k++;
            }
            else{
                arr[k]=rightArr[j];
                j++;
                k++;
            }
        }

        while(i<nl){
            arr[k]=leftArr[i];
            i++;
            k++;
        }
        while(j<nr){
            arr[k]=rightArr[j];
            j++;
            k++;
        }

    }
    static void mergeSort(int [] arr, int beg , int end )
    {
        if(beg<end)
        {
            int mid = (beg + end) /2;
            mergeSort(arr,beg,mid);
            mergeSort(arr,mid+1,end);
            merge(arr,beg,mid,end);
        }
    }
    
    static void print(int arr[],int n)
    {

        for (int i = 0; i < n; i++) {
            System.out.print(arr[i]+" ");
        }
        System.out.println();
    }
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter the number of elements: ");
        int n = sc.nextInt();
        int arr[] = new int [n];
        System.out.print("Enter elements : ");
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }
        System.out.print("Array before sort: ");
        print(arr,n);
        mergeSort(arr,0,n-1);
        System.out.print("Array after sort: ");
        print(arr,n);

    }
}
